File: Java/Question Practice Java/Circular_Linked_List.java
import java.util.Scanner;

class Node {
    int data;
    Node next;

    public Node(int data) {
        this.data = data;
        this.next = null;
    }
}

class CircularLinkedList {
    Node head;

    public CircularLinkedList() {this.head = null;}

    public void InsertAtBeginning(int val) {
        Node newNode = new Node(val);
        if (head == null) {
            head = newNode;
            newNode.next = head;
        } else {
            Node temp = head;
            while (temp.next != head) temp = temp.next;
            newNode.next = head;
            temp.next = newNode;
            head = newNode;
        } System.out.println("\nInserted head node with value " + newNode.data);
    }

    public void InsertAtEnd(int val) {
        if (head == null) {
            InsertAtBeginning(val);
            return;
        }
        Node newNode = new Node(val);
        Node temp = head;
        while (temp.next != head) temp = temp.next;
        temp.next = newNode;
        newNode.next = head;
        System.out.println("\nInserted new node with value " + newNode.data);
    }

    public void DeleteAtBeginning() {
        if (head == null) {
            System.out.println("\nList is empty");
            return;
        }
        if (head.next == head) {
            System.out.println("\nDeleted node with value " + head.data);
            head = null;
            return;
        }
        Node temp = head;
        Node last = head;
        while (last.next != head) last = last.next;
        head = head.next;
        last.next = head;
        System.out.println("\nDeleted node with value " + temp.data);
    }

    public void DeleteAtEnd() {
        if (head == null) {
            System.out.println("\nList is empty");
            return;
        }
        if (head.next == head) {
            System.out.println("\nDeleted node with value " + head.data);
            head = null;
            return;
        }
        Node temp = head;
        while (temp.next.next != head) temp = temp.next;
        System.out.println("\nDeleted node with value " + temp.next.data);
        temp.next = head;
    }

    public void Display() {
        if (head == null) {
            System.out.println("\nList is empty");
            return;
        }
        Node temp = head;
        System.out.println();
        do {
            System.out.print(temp.data + " -> ");
            temp = temp.next;
        } while (temp != head);
        System.out.println("(head)");
    }
}

public class Circular_Linked_List {
    public static void main(String[] args) {
        Scanner input = new Scanner(System.in);
        CircularLinkedList list = new CircularLinkedList();

        System.out.println("Circular Linked List Operations:\n1. InsertAtBeginning\n2. InsertAtEnd\n3. DeleteAtBeginning\n4. DeleteAtEnd\n5. Display");
        while (true) {
            System.out.print("\nEnter Linked List Function: ");
            int choice = input.nextInt();

            if (choice == 1) {
                System.out.print("Enter the Element: ");
                int elem = input.nextInt();
                list.InsertAtBeginning(elem);
            } else if (choice == 2) {
                System.out.print("Enter the Element: ");
                int elem = input.nextInt();
                list.InsertAtEnd(elem);
            } else if (choice == 3) list.DeleteAtBeginning();
            else if (choice == 4) list.DeleteAtEnd();
            else if (choice == 5) list.Display();
            else System.out.println("Invalid Choice");
        }
    }
}

File: C/Question Practice C/Circular_Linked_List.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node* next;
};

struct LinkedList {struct Node* head;};

void Initialize(struct LinkedList* list) {list->head = NULL;}

void InsertAtBeginning(struct LinkedList* list, int val) {
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode->data = val;

    if (list->head == NULL) {
        newNode->next = newNode;
        list->head = newNode;
    } else {
        struct Node* temp = list->head;
        while (temp->next != list->head) temp = temp->next;

        newNode->next = list->head;
        temp->next = newNode;
        list->head = newNode;
    }

    printf("Inserted head node with value %d\n", newNode->data);
}

void InsertAtEnd(struct LinkedList* list, int val) {
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode->data = val;

    if (list->head == NULL) {
        newNode->next = newNode;
        list->head = newNode;
        printf("Inserted head node with value %d\n", newNode->data);
        return;
    }

    struct Node* temp = list->head;
    while (temp->next != list->head) temp = temp->next;

    temp->next = newNode;
    newNode->next = list->head;

    printf("Inserted new node with value %d\n", newNode->data);
}

void DeleteAtBeginning(struct LinkedList* list) {
    if (list->head == NULL) {
        printf("List is empty\n");
        return;
    }

    struct Node* temp = list->head;

    if (temp->next == list->head) {
        printf("Deleted head node with value %d\n", temp->data);
        free(temp);
        list->head = NULL;
        return;
    }

    struct Node* last = list->head;
    while (last->next != list->head) last = last->next;

    list->head = list->head->next;
    last->next = list->head;

    printf("Deleted head node with value %d\n", temp->data);
    free(temp);
}

void DeleteAtEnd(struct LinkedList* list) {
    if (list->head == NULL) {
        printf("List is empty\n");
        return;
    }

    struct Node* temp = list->head;

    if (temp->next == list->head) {
        printf("Deleted head node with value %d\n", temp->data);
        free(temp);
        list->head = NULL;
        return;
    }

    struct Node* prev = NULL;
    do{
        prev = temp;
        temp = temp->next;
    } while (temp->next != list->head);

    prev->next = list->head;
    printf("Deleted node with value %d\n", temp->data);
    free(temp);
}

void Display(struct LinkedList* list) {
    if (list->head == NULL) {
        printf("List is empty\n");
        return;
    }

    struct Node* temp = list->head;
    do {
        printf("%d -> ", temp->data);
        temp = temp->next;
    } while (temp != list->head);
    printf("(head)\n");
}

int main() {
    struct LinkedList l1;
    Initialize(&l1);

    printf("Circular Singly Linked List Operations:\n");
    printf("1. InsertAtBeginning\n2. InsertAtEnd\n3. DeleteAtBeginning\n4. DeleteAtEnd\n5. Display\n");

    while (1) {
        int choice, elem;
        printf("\nEnter Linked List Operation: ");
        scanf("%d", &choice);

        if (choice == 1) {
            printf("Enter the Value to insert: ");
            scanf("%d", &elem);
            InsertAtBeginning(&l1, elem);
        } else if (choice == 2) {
            printf("Enter the Value to insert: ");
            scanf("%d", &elem);
            InsertAtEnd(&l1, elem);
        } 
        else if (choice == 3) DeleteAtBeginning(&l1);
        else if (choice == 4) DeleteAtEnd(&l1);
        else if (choice == 5) Display(&l1);
        else printf("Enter Valid Operation\n");
    }

    return 0;
}